http/listener: fix bugs and simplify

denpeshkov opened the pull request:
## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license](https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Description
1. Store `ctx.Done` channel in a struct instead of a `ctx`. See: https://go.dev/blog/context-and-structs
2. Return from `handleListener` on `ctx` cancellation preventing goroutines leaks
3. Simplify `handleListener` by removing `send` closure. The `handleListener` is inlined by the compiler
4. Return first error from `Close`
5. Preallocate slice in `Addrs`
6. Reduce duplication in handling `opts.Trace`
7. Prevent original listener address from being modified in `Addr` call
8. Clip unused elements from listeners slice in the constructor

## Motivation and Context
Fix bugs and simplify code

## How to test this PR?


## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

denpeshkov committed:
http/listener: fix bugs and simplify

1. Store `ctx.Done` channel in a struct instead of a `ctx`. See: https://go.dev/blog/context-and-structs
2. Return from `handleListener` on `ctx` cancellation preventing goroutines leaks
3. Simplify `handleListener` by removing `send` closure. The `handleListener` is inlined by the compiler
4. Return first error from `Close`
5. Preallocate slice in `Addrs`
6. Reduce duplication in handling `opts.Trace`

klauspost commented on internal/http/listener.go:
```diff
@@ -87,10 +80,12 @@ func (listener *httpListener) Close() (err error) {
 	for i := range listener.listeners {
-		listener.listeners[i].Close()
+		if cerr := listener.listeners[i].Close(); cerr != nil && err == nil {
+			err = cerr
```
Errors don't really provide much value at this point and will just give noisy logs on shutdown. I suggest you revert this change.

denpeshkov replied:
Reverted

klauspost reviewed (commented):
Nice cleanup.

denpeshkov committed:
http/listener: revert error propagation from Close()

klauspost reviewed (approved):
lgtm

denpeshkov committed:
http/listener: preserve original listener address in Addr()

- Preserve the original address when calling Addr() with multiple listeners
- Remove unused listeners from the slice
